roles/conduit/files/conduit_leds.py

```diff
@@ -0,0 +1,507 @@
+
+def check_tunnel(options):
+    """ Check tunnel status """
+
+    global cached_ip
+
+    try:
+        config = Defaults("ssh_tunnel")
+    except IOError:
+        cached_ip = None
+        return False
+
+    local_port = config.get('LOCAL_PORT')
+    if not local_port:
+        cached_ip = None
+        return False
+    remote_host = config.get('REMOTE_HOST')
+    if not remote_host:
+        cached_ip = None
+        return False
+    try:
+        remote_ip = socket.gethostbyname(remote_host)
+        cached_ip = remote_ip
+    except socket.gaierror:
+        if not cached_ip:
+            logging.info("check_tunnel: Unable to resolve %s", remote_host)
+            return False
+        remote_host = cached_ip
+        logging.info("check_tunnel: Using cached IP %s", remote_host)
+
+    for conn in psutil.net_connections():
+        if conn.type == socket.SOCK_STREAM and conn.status == psutil.CONN_ESTABLISHED and conn.raddr == (remote_ip, local_port):
+            logging.info("check_tunnel: Found connection to %s(%s):%s with PID %d",
+                         remote_host,
+                         remote_ip,
+                         local_port,
+                         conn.pid)
+            return True
+
+    logging.info("check_tunnel: No connection found to %s(%s):%s", remote_host, remote_ip, local_port)
+    return False
+
+def check_dns(options):
+    """ Check dns status """
+
+    try:
+        host_ip = socket.gethostbyname(options.hostname)
+        logging.info("check_dns: %s resolved to %s", options.hostname, host_ip)
+    except socket.gaierror:
+        logging.info("check_dns: %s failed to resolve", options.hostname)
+        return False
+
+    return True
+
+def check_lora(options, device_path):
+    """ Check LoRa status """
+
+    cmd = ["fuser", device_path]
+    DEVNULL = open(os.devnull, 'w')
+    try:
+        subprocess.check_call(cmd,
+                              stdout=DEVNULL,
+                              stderr=subprocess.STDOUT)
+        logging.info("check_lora: fuser %s returned 0", device_path)
+    except subprocess.CalledProcessError:
+        logging.info("check_lora: fuser %s returned non-zero", device_path)
+        return False
+
+    return True
+
+def check_ppp(options):
+    """ Check status of ppp connection """
+
+    try:
+        modem_stat = os.stat(options.modem)
+        if not stat.S_ISCHR(modem_stat.st_mode):
+            logging.debug("check_ppp: %s not a character device", options.modem)
+            return False
+    except OSError as error:
+        logging.debug("check_ppp: %s: %s", options.modem, error)
+        return False
+
+    peer_addr = None
+    for ifname, ifaddrs in psutil.net_if_addrs().items():
+        if ifname != "ppp0":
+            continue
+        for ifaddr in ifaddrs:
+            if ifaddr.family != socket.AF_INET:
+                continue
+            if ifaddr.ptp is None:
+                continue
+            if ifaddr.ptp == "10.64.64.64":
```

Comment:
what's this magic number?

Reply:
pppd's `hisaddr`.  10.64.64.64 + ppp index (ppp0 = 0) or 10.112.112.112 for dial on demand.  This means the remote has not yet given us their address.  I find that the connection isn't working at that point. I improved the code to properly detect both types of addresses including some documentation.